server/app/api/v1/charity/raffle/controller.js

```diff
@@ -0,0 +1,219 @@
+ */
+const all = (req, res, next) => {
+  if (req.query.hasOwnProperty('isActive') && req.query.isActive === 'true') {
+    return Raffle.find({ isActive: true })
+      .select('-charity')
+      .populate('winner', 'email firstName lastName')
+      .exec()
+      .then(raffles => res.json(raffles))
+      .catch(err => next(err));
+  }
+  return Charity
+    .populate(req.charity, {
+      path: 'raffles',
+      select: '-charity'
+    })
+    .then(({ raffles }) => {
+      return Raffle.populate(raffles, {
+        path: 'winner'
+      });
+    })
+    .then(raffles => res.json(raffles))
+    .catch(err => next(err));
+};
+
+/**
+ * Returns req.raffle
+ */
+const get = (req, res) => res.json(req.raffle);
+
+/**
+ * Creates new raffle
+ */
+const create = (req, res, next) => {
+  return new Raffle({
+    charity: req.charity._id,
+    prize: {
+      name: req.body.prizeName,
+      thumbnailURL: req.body.prizeThumbnailURL
+    }
+  }).save()
+  .then((savedRaffle) => {
+    req.charity.raffles.push(savedRaffle);
+    req.charity.save();
+    savedRaffle = savedRaffle.toObject();
+    return res.json(savedRaffle);
+  })
+  .catch(err => next(err));
+};
+
+/**
+ * Purchase raffle
+ * if existing user use query userId={userId}
+ * if new user user email, firstName, lastName in req.body
+ */
+
+const purchase = (req, res, next) => {
+  const hasUserId = req.query.hasOwnProperty('userId');
+  if (req.raffle.isActive) {
+    const findOrCreateUser = hasUserId ? findByUserId : createUser;
+
+    findOrCreateUser()
+      .then(addTickets)
+        .all()
+      .then(([raffle, user]) => {
+        const { email, firstName, lastName } = user;
+        const { count } = raffle.purchases.filter(
+          purchase => purchase.buyer === user._id
+        )[0];
+        return res.json({
+          count,
+          user: {
+            email,
+            firstName,
+            lastName
+          }
+        });
+      })
+      .catch(err => next(err));
+  } else {
+    next(
+      new NotPermittedError(
+        'Tickets cannot be purchased after raffle has ended.'
+      )
+    );
+  }
+
+  function findByUserId() {
+    const userId = req.query.userId;
+
+    if (!userId.match(VALID_OBJECT_ID)) {
+      return next(
+        new ValidationError(
+          'charityId must be valid ObjectId',
+          VALID_OBJECT_ID,
+          'charityId'
+        )
+      );
+    }
+    return User.findById(userId)
+      .then((user) => {
+        if (!user) {
+          throw new NotFoundError('User not found');
+        } else {
+          return user;
+        }
+      });
+  }
+
+  function createUser() {
+    return new User({
+      firstName: req.body.firstName,
+      lastName: req.body.lastName,
+      email: req.body.email
+    }).save();
+  }
+
+  function addTickets(user) {
+    // checks for if user already purchased raffles
+    const purchaseIndex = findIndex(req.raffle.purchases, ['buyer', user._id]);
```

Comment:
See the `Raffle` [model schema](https://github.com/lackeyjb/velociraffle-web-app/blob/feature/raffles/server/app/api/v1/charity/raffle/model.js#L37). Each `Raffle` has an array of `purchases` (where a purchase has a `count` and a `buyer`). I'm updating the `count` if the `buyer` (user) has already purchased tickets and wants to buy more. `purchases` might not be the best word to describe this array, @rgpass do you have any ideas?